File: order-service/src/main/java/com/ecommerce/orderservice/config/GlobalExceptionHandler.java
package com.ecommerce.orderservice.config;

import java.time.LocalDateTime;
import java.util.Map;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;

@ControllerAdvice
public class GlobalExceptionHandler {

  @ExceptionHandler(IllegalArgumentException.class)
  public ResponseEntity<Object> handleIllegalArgumentException(IllegalArgumentException ex) {
    return new ResponseEntity<>(
        Map.of(
            "timestamp", LocalDateTime.now(),
            "status", HttpStatus.BAD_REQUEST.value(),
            "error", "Bad Request",
            "message", ex.getMessage()
        ),
        HttpStatus.BAD_REQUEST
    );
  }

  // Optionally catch generic exceptions too
  @ExceptionHandler(Exception.class)
  public ResponseEntity<Object> handleAllExceptions(Exception ex) {
    return new ResponseEntity<>(
        Map.of(
            "timestamp", LocalDateTime.now(),
            "status", HttpStatus.INTERNAL_SERVER_ERROR.value(),
            "error", "Internal Server Error",
            "message", "An unexpected error occurred."
        ),
        HttpStatus.INTERNAL_SERVER_ERROR
    );
  }
}

File: user-service/src/main/kotlin/com/ecommerce/userservice/config/GlobalExceptionHandler.kt
package com.ecommerce.userservice.config

import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.ExceptionHandler
import org.springframework.web.bind.annotation.RestControllerAdvice


@RestControllerAdvice
class GlobalExceptionHandler {

    @ExceptionHandler(Exception::class)
    fun handleUserExists(ex: Exception): ResponseEntity<MutableMap<String?, String?>?> {
        val error: MutableMap<String?, String?> = HashMap<String?, String?>()
        error.put("error", ex.message)
        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
            .body<MutableMap<String?, String?>?>(error)
    }
}